File: backend/app/schemas/kpi_task.py
"""
KPI Task Schemas

Pydantic models for KPI task management API.
"""

from typing import Optional
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field, field_validator

from app.db.models import KPITaskStatusEnum, KPITaskPriorityEnum


# Base schema with common fields
class KPITaskBase(BaseModel):
    """Base schema for KPI Task"""
    title: str = Field(..., min_length=1, max_length=255, description="Название задачи")
    description: Optional[str] = Field(None, description="Подробное описание задачи")
    status: KPITaskStatusEnum = Field(
        default=KPITaskStatusEnum.TODO,
        description="Статус задачи"
    )
    priority: KPITaskPriorityEnum = Field(
        default=KPITaskPriorityEnum.MEDIUM,
        description="Приоритет задачи"
    )
    complexity: Optional[int] = Field(
        None,
        ge=1,
        le=10,
        description="Сложность задачи по шкале 1-10"
    )
    estimated_hours: Optional[Decimal] = Field(
        None,
        ge=0,
        description="Оценка времени выполнения в часах"
    )
    actual_hours: Optional[Decimal] = Field(
        None,
        ge=0,
        description="Фактическое время выполнения в часах"
    )
    completion_percentage: Optional[Decimal] = Field(
        default=Decimal("0"),
        ge=0,
        le=100,
        description="Процент выполнения задачи"
    )
    due_date: Optional[datetime] = Field(None, description="Срок выполнения")
    start_date: Optional[datetime] = Field(None, description="Дата начала работы")
    notes: Optional[str] = Field(None, description="Комментарии и заметки")


class KPITaskCreate(KPITaskBase):
    """Schema for creating a new KPI Task"""
    employee_kpi_goal_id: int = Field(..., description="ID цели KPI")
    employee_id: int = Field(..., description="ID сотрудника")
    # department_id берется из current_user, assigned_by_id тоже

    @field_validator('due_date', 'start_date')
    @classmethod
    def validate_dates(cls, v):
        """Validate that dates are not in the past (optional)"""
        # Optional: можно добавить валидацию, что due_date >= start_date
        return v


class KPITaskUpdate(BaseModel):
    """Schema for updating an existing KPI Task"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    status: Optional[KPITaskStatusEnum] = None
    priority: Optional[KPITaskPriorityEnum] = None
    complexity: Optional[int] = Field(None, ge=1, le=10)
    estimated_hours: Optional[Decimal] = Field(None, ge=0)
    actual_hours: Optional[Decimal] = Field(None, ge=0)
    completion_percentage: Optional[Decimal] = Field(None, ge=0, le=100)
    due_date: Optional[datetime] = None
    start_date: Optional[datetime] = None
    notes: Optional[str] = None


class KPITaskInDB(KPITaskBase):
    """Schema for KPI Task from database"""
    id: int
    employee_kpi_goal_id: int
    employee_id: int
    assigned_by_id: Optional[int]
    department_id: int
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class KPITaskResponse(KPITaskInDB):
    """Schema for KPI Task API response with related data"""
    # Optional: можно добавить связанные данные
    employee_name: Optional[str] = None
    goal_name: Optional[str] = None
    assigned_by_name: Optional[str] = None


class KPITaskStatusUpdate(BaseModel):
    """Schema for updating task status"""
    status: KPITaskStatusEnum
    completion_percentage: Optional[Decimal] = Field(None, ge=0, le=100)
    notes: Optional[str] = None

    @field_validator('completion_percentage')
    @classmethod
    def validate_completion(cls, v, info):
        """Auto-set completion based on status"""
        status = info.data.get('status')
        if status == KPITaskStatusEnum.DONE and v is None:
            return Decimal("100")
        if status == KPITaskStatusEnum.TODO and v is None:
            return Decimal("0")
        return v


class KPITaskComplexityUpdate(BaseModel):
    """Schema for updating task complexity and estimation"""
    complexity: int = Field(..., ge=1, le=10)
    estimated_hours: Optional[Decimal] = Field(None, ge=0)
    notes: Optional[str] = None


class KPITaskBulkStatusUpdate(BaseModel):
    """Schema for bulk status update"""
    task_ids: list[int] = Field(..., min_length=1, description="Список ID задач")
    status: KPITaskStatusEnum
    notes: Optional[str] = None


class KPITaskStatistics(BaseModel):
    """Schema for task statistics"""
    total_tasks: int
    by_status: dict[str, int]
    by_priority: dict[str, int]
    avg_complexity: Optional[Decimal]
    total_estimated_hours: Optional[Decimal]
    total_actual_hours: Optional[Decimal]
    completion_rate: Decimal  # % задач в статусе DONE
    overdue_tasks: int

File: frontend/src/types/kpiTask.ts
/**
 * KPI Task Types
 *
 * TypeScript type definitions for KPI task management.
 */

// Enums matching backend
export type KPITaskStatus = 'TODO' | 'IN_PROGRESS' | 'DONE' | 'CANCELLED'

export type KPITaskPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'

// Core task interface
export interface KPITask {
  id: number
  employee_kpi_goal_id: number
  employee_id: number
  assigned_by_id: number | null
  department_id: number

  // Task details
  title: string
  description: string | null
  status: KPITaskStatus
  priority: KPITaskPriority

  // Complexity and estimation
  complexity: number | null // 1-10 scale
  estimated_hours: number | null
  actual_hours: number | null

  // Progress
  completion_percentage: number | null // 0-100%

  // Dates
  due_date: string | null
  start_date: string | null
  completed_at: string | null

  // Additional
  notes: string | null

  // Timestamps
  created_at: string
  updated_at: string
}

// Extended response with related data
export interface KPITaskResponse extends KPITask {
  employee_name?: string | null
  goal_name?: string | null
  assigned_by_name?: string | null
}

// Create task request
export interface KPITaskCreate {
  employee_kpi_goal_id: number
  employee_id: number
  title: string
  description?: string | null
  status?: KPITaskStatus
  priority?: KPITaskPriority
  complexity?: number | null
  estimated_hours?: number | null
  due_date?: string | null
  start_date?: string | null
  notes?: string | null
}

// Update task request
export interface KPITaskUpdate {
  title?: string
  description?: string | null
  status?: KPITaskStatus
  priority?: KPITaskPriority
  complexity?: number | null
  estimated_hours?: number | null
  actual_hours?: number | null
  completion_percentage?: number | null
  due_date?: string | null
  start_date?: string | null
  notes?: string | null
}

// Status update request
export interface KPITaskStatusUpdate {
  status: KPITaskStatus
  completion_percentage?: number | null
  notes?: string | null
}

// Complexity update request
export interface KPITaskComplexityUpdate {
  complexity: number // 1-10
  estimated_hours?: number | null
  notes?: string | null
}

// Bulk status update
export interface KPITaskBulkStatusUpdate {
  task_ids: number[]
  status: KPITaskStatus
  notes?: string | null
}

// Statistics
export interface KPITaskStatistics {
  total_tasks: number
  by_status: Record<KPITaskStatus, number>
  by_priority: Record<KPITaskPriority, number>
  avg_complexity: number | null
  total_estimated_hours: number | null
  total_actual_hours: number | null
  completion_rate: number // percentage
  overdue_tasks: number
}

// Filter options for API requests
export interface KPITaskFilters {
  employee_id?: number
  employee_kpi_goal_id?: number
  status?: KPITaskStatus
  priority?: KPITaskPriority
  overdue_only?: boolean
  department_id?: number
  skip?: number
  limit?: number
}

// UI-specific types
export interface KPITaskFormValues {
  title: string
  description: string
  status: KPITaskStatus
  priority: KPITaskPriority
  complexity: number | null
  estimated_hours: number | null
  due_date: string | null
  notes: string
}

// Task priority display config
export const TASK_PRIORITY_CONFIG: Record<
  KPITaskPriority,
  { label: string; color: string; badge: string }
> = {
  LOW: { label: 'Низкий', color: 'default', badge: 'L' },
  MEDIUM: { label: 'Средний', color: 'blue', badge: 'M' },
  HIGH: { label: 'Высокий', color: 'orange', badge: 'H' },
  CRITICAL: { label: 'Критический', color: 'red', badge: '!' },
}

// Task status display config
export const TASK_STATUS_CONFIG: Record<
  KPITaskStatus,
  { label: string; color: string; icon: string }
> = {
  TODO: { label: 'К выполнению', color: 'default', icon: 'ClockCircleOutlined' },
  IN_PROGRESS: { label: 'В работе', color: 'processing', icon: 'SyncOutlined' },
  DONE: { label: 'Выполнено', color: 'success', icon: 'CheckCircleOutlined' },
  CANCELLED: { label: 'Отменено', color: 'error', icon: 'CloseCircleOutlined' },
}

// Complexity scale labels
export const COMPLEXITY_LABELS: Record<number, string> = {
  1: 'Очень простая',
  2: 'Простая',
  3: 'Несложная',
  4: 'Средняя',
  5: 'Умеренная',
  6: 'Умеренно сложная',
  7: 'Сложная',
  8: 'Очень сложная',
  9: 'Крайне сложная',
  10: 'Максимальная сложность',
}
